3b.run_sam/scripts/1.run_sam2_microscopy.py

```diff
@@ -0,0 +1,615 @@
+        False  # set True to generate negative samples for better training
+    )
+    # loop through the points and add them to the state and get the masks
+    for _point, _label in zip(
+        image_set_dict["image_x_y_coords"][i], image_set_dict["image_labels"][i]
+    ):
+        _label = np.array([_label], dtype=np.int32)
+        _point = np.array([_point], dtype=np.float32)
+
+        if negative_sampling:
+            random_points, random_labels = generate_random_coords(
+                img=img, coords=_point, samples=samples
+            )
+            _point = np.concatenate([_point, random_points], axis=0)
+            _label = np.concatenate([_label, random_labels], axis=0)
+        # add the points to the state
+        _, out_obj_ids, out_mask_logits = predictor.add_new_points(
+            inference_state=inference_state,
+            frame_idx=ann_frame_idx,
+            obj_id=ann_obj_idx,
+            points=_point,
+            labels=_label,
+        )
+        # save the prompts
+        prompts[ann_obj_idx] = {
+            "points": _point,
+            "labels": _label,
+            "out_obj_ids": out_obj_ids[0],
+            "out_mask_logits": out_mask_logits[0].detach().cpu().numpy(),
+        }
+        # increment the object index for this frame
+        ann_obj_idx += 1
+
+    del prompts
+    del samples
+    # run propagation throughout the video and collect the results in a dict
+    video_segments = {}  # video_segments contains the per-frame segmentation results
+
+    for out_frame_idx, out_obj_ids, out_mask_logits in predictor.propagate_in_video(
+        inference_state
+    ):
+        video_segments[out_frame_idx] = {
+            out_obj_id: (out_mask_logits[i] > 0.0).cpu().numpy()
+            for i, out_obj_id in enumerate(range(1, ann_obj_idx))
+        }
+    stored_video_segments[image_set_dict["image_set_name"][i]] = video_segments
+
+    # clear the memory
+    del inference_state
+
+    del out_mask_logits
+    del out_obj_ids
+    del out_frame_idx
+    torch.cuda.empty_cache()
+    gc.collect()
+
+
+# ### stop GPU profiling
+
+# In[ ]:
+
+
+# save the memory snapshot to a file
+export_memory_snapshot(
+    logger=logger, save_dir=pathlib.Path("../memory_snapshots/").resolve()
+)
+stop_record_memory_history(logger=logger)
+
+
+# In[ ]:
+
+
+masks_dir = pathlib.Path("../sam2_processing_dir/masks").resolve()
+if masks_dir.exists():
+    shutil.rmtree(masks_dir)
+masks_dir.mkdir(exist_ok=True, parents=True)
+gifs_dir = pathlib.Path("../sam2_processing_dir/gifs").resolve()
+if gifs_dir.exists():
+    shutil.rmtree(gifs_dir)
+gifs_dir.mkdir(exist_ok=True, parents=True)
+
+
+# In[ ]:
+
+
+output_dict = {
+    "image_set_name": [],
+    "frame": [],
+    "object_id": [],
+    "x": [],
+    "y": [],
+    "mask_path": [],
+    "mask_file_name": [],
+}
+
+
+# In[ ]:
+
+
+# loop through each image set and predict the instances
+for i in range(len(image_set_dict["image_set_name"])):
```

Comment:
This is the same code comment for the loop above. What is the difference and why have two loops unless they are doing different things. Recommend updating this code comment.